core/shared/src/main/scala/org/typelevel/log4cats/extras/DeferredLoggerFactory.scala

```diff
@@ -0,0 +1,150 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats.extras
+
+import cats.Show.Shown
+import cats.data.Chain
+import cats.effect.kernel.{Concurrent, Resource}
+import cats.syntax.all.*
+import cats.{~>, Functor}
+import org.typelevel.log4cats.extras.DeferredStructuredLogger.DeferredStructuredLogMessage
+import org.typelevel.log4cats.{LoggerFactory, SelfAwareStructuredLogger}
+
+/**
+ * A `LoggerFactory` that does not immediately log.
+ *
+ * Effectively a `LoggerFactory` equivalent to `DeferredSelfAwareStructuredLogger`. As an
+ * implementation note, the `LoggerFactory` trait is constrained in such a way that this will
+ * produce `SelfAwareStructuredLogger`, rather than `DeferredSelfAwareStructuredLogger`, so if
+ * logging is desired it needs to be triggered using the `DeferredLoggerFactory`, rather than being
+ * able to trigger it from any of the produced loggers.
+ *
+ * >>> WARNING: READ BEFORE USAGE! <<<
+ * https://github.com/typelevel/log4cats/blob/main/core/shared/src/main/scala/org/typelevel/log4cats/extras/README.md
+ * >>> WARNING: READ BEFORE USAGE! <<<
+ */
+trait DeferredLoggerFactory[F[_]] extends LoggerFactory[F] {
+
+  /**
+   * View the logs in the buffer.
+   *
+   * This is primarily useful for testing, and will not effect the behavior of calls to `log`
+   */
+  def inspect: F[Chain[DeferredStructuredLogMessage]]
+
+  /**
+   * Log the deferred messages
+   *
+   * This may be called multiple times, and each log should only be logged once.
+   */
+  def log: F[Unit]
+
+  override def getLoggerFromName(name: String): SelfAwareStructuredLogger[F]
+
+  override def addContext(ctx: Map[String, String])(implicit
+      F: Functor[F]
+  ): DeferredLoggerFactory[F] =
+    DeferredLoggerFactory.addContext(this, ctx)
+
+  override def addContext(pairs: (String, Shown)*)(implicit
+      F: Functor[F]
+  ): DeferredLoggerFactory[F] =
+    DeferredLoggerFactory.addContext(this, pairs.map { case (k, v) => (k, v.toString) }.toMap)
+
+  override def withModifiedString(f: String => String)(implicit
+      F: Functor[F]
+  ): DeferredLoggerFactory[F] =
+    DeferredLoggerFactory.withModifiedString(this, f)
+
+  override def mapK[G[_]](fk: F ~> G)(implicit F: Functor[F]): DeferredLoggerFactory[G] =
+    DeferredLoggerFactory.mapK[F, G](fk)(this)
+}
+object DeferredLoggerFactory {
+
+  def apply[F[_]: Concurrent](
+      loggerFactory: LoggerFactory[F]
+  ): Resource[F, DeferredLoggerFactory[F]] =
+    DeferredSelfAwareStructuredLogger.makeCache[F].map { cache =>
```

Comment:
This ended up paralleling a similar dependency in `LoggerFactory`, as that conceptually depends on `SelfAwareStructuredLogger` (it's a bit hard to see in the code, but `LoggerType` is a big clue).

We need `DeferredLoggerFactory` because, the way `LoggerFactory` is written, it's near enough impossible to create a `LoggerFactory` that emits `DeferredSelfAwareStructuredLogger`s that are aware that they are `DeferredSelfAwareStructuredLogger`s rather than vanilla `SelfAwareStructuredLogger`s - which then makes it impossible to manually trigger logging for errors that don't involve an error embedded in `F[_]`